File: ios/Player/PlaybackIntervalManager.swift
import Foundation
#if canImport(NitroModules)
import NitroModules
#endif

/// Emits a fixed-cadence tick while playback is active AND enabled.
/// Carries no payload; consumers measure playback time themselves.
@MainActor
class PlaybackIntervalManager {
  private static let tickInterval: TimeInterval = 1.0

  nonisolated(unsafe) private var timer: Timer?
  private var enabled = false
  private var active = false
  private let onTick: () -> Void

  init(onTick: @escaping () -> Void) {
    self.onTick = onTick
  }

  deinit {
    timer?.invalidate()
  }

  func setEnabled(_ enabled: Bool) {
    if self.enabled == enabled { return }
    self.enabled = enabled
    reconcile()
  }

  func onPlaybackStateChanged(_ state: PlaybackState) {
    switch state {
    // Only actual playback counts as "playback time"; buffering/loading freeze
    // the clock so consumers measure cumulative *playing* time.
    case .playing:
      active = true
    case .loading, .buffering, .paused, .stopped, .ended, .error:
      active = false
    default:
      break
    }
    reconcile()
  }

  private func reconcile() {
    if enabled && active {
      start()
    } else {
      stop()
    }
  }

  private func start() {
    guard timer == nil else { return }
    timer = Timer.scheduledTimer(
      withTimeInterval: Self.tickInterval,
      repeats: true
    ) { [weak self] _ in
      MainActor.assumeIsolated {
        self?.onTick()
      }
    }
  }

  private func stop() {
    timer?.invalidate()
    timer = nil
  }
}

File: android/src/main/java/com/audiobrowser/player/PlaybackIntervalManager.kt
package com.audiobrowser.player

import com.margelo.nitro.audiobrowser.PlaybackState
import kotlinx.coroutines.Job
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/** Emits a fixed-cadence tick while playback is active AND enabled. */
class PlaybackIntervalManager(private val onTick: () -> Unit) {
  private val scope = MainScope()
  private var job: Job? = null
  private var enabled = false
  private var active = false

  fun setEnabled(value: Boolean) {
    if (enabled == value) return
    enabled = value
    reconcile()
  }

  fun onPlaybackStateChanged(state: PlaybackState) {
    active =
      when (state) {
        // Only actual playback counts as "playback time"; buffering/loading
        // freeze the clock so consumers measure cumulative *playing* time.
        PlaybackState.PLAYING -> true
        PlaybackState.LOADING,
        PlaybackState.BUFFERING,
        PlaybackState.PAUSED,
        PlaybackState.STOPPED,
        PlaybackState.ENDED,
        PlaybackState.ERROR -> false
        else -> active
      }
    reconcile()
  }

  private fun reconcile() {
    if (enabled && active) start() else stop()
  }

  private fun start() {
    if (job != null) return
    job =
      scope.launch {
        while (true) {
          delay(TICK_MS)
          onTick()
        }
      }
  }

  private fun stop() {
    job?.cancel()
    job = null
  }

  companion object {
    private const val TICK_MS = 1000L
  }
}
